src/verifier/snp/mod.rs

```diff
@@ -0,0 +1,141 @@
+use anyhow::{anyhow, Context, Result};
+extern crate serde;
+use self::serde::{Deserialize, Serialize};
+use super::*;
+use async_trait::async_trait;
+use serde_json::json;
+use sev::firmware::guest::types::AttestationReport;
+use sev::firmware::host::types::{CertTable, SnpCertType};
+
+use openssl::{
+    ec::EcKey,
+    ecdsa,
+    pkey::{PKey, Public},
+    x509,
+};
+use sha2::{Digest, Sha384};
+
+#[derive(Serialize, Deserialize)]
+struct SnpEvidence {
+    attestation_report: AttestationReport,
+    cert_chain: CertTable,
+}
+
+#[derive(Debug, Default)]
+pub struct Snp {}
+
+#[async_trait]
+impl Verifier for Snp {
+    async fn evaluate(
+        &self,
+        nonce: String,
+        attestation: &Attestation,
+    ) -> Result<TeeEvidenceParsedClaim> {
+        let tee_evidence = serde_json::from_str::<SnpEvidence>(&attestation.tee_evidence)
+            .context("Deserialize Quote failed.")?;
+
+        verify_report_signature(&tee_evidence)?;
+
+        let report = tee_evidence.attestation_report;
+        if report.version != 2 {
+            return Err(anyhow!("Unexpected report version"));
+        }
+
+        if report.vmpl != 0 {
+            return Err(anyhow!("VMPL Check Failed"));
+        }
+
+        let expected_report_data = calculate_expected_report_data(&nonce, &attestation.tee_pubkey);
+        if report.report_data != expected_report_data {
+            return Err(anyhow!("Report Data Mismatch"));
+        }
+
+        parse_tee_evidence(&report)
+    }
+}
+
+fn verify_report_signature(evidence: &SnpEvidence) -> Result<()> {
+    // check cert chain
+    let vcek = verify_cert_chain(&evidence.cert_chain)?;
+
+    // verify report signature
+    let sig = ecdsa::EcdsaSig::try_from(&evidence.attestation_report.signature)?;
+    let data = &bincode::serialize(&evidence.attestation_report)?[..=0x29];
+
+    sig.verify(&data, EcKey::try_from(vcek.public_key()?)?.as_ref())
+        .map_err(|_| anyhow!("Signature validation failed."))?;
+
+    Ok(())
+}
+
+fn verify_cert_chain(cert_chain: &CertTable) -> Result<x509::X509> {
+    // add support for genoa
+    let certs = x509::X509::stack_from_pem(include_bytes!("milan_ark_ask.pem"))?;
```

Comment:
Given the rare possibility of a key revocation, I am assuming we would just update the file locally?